src/coreclr/jit/compiler.h

```diff
@@ -10933,6 +10933,13 @@ class Compiler
 
     bool compObjectStackAllocation()
     {
+        if (compIsAsync2())
```

Comment:
In theory we could treat async calls as "escaping" object refs that are live after a call. That way temps used in between cc calls could still be stack allocated. 
Probably not the high priority though.

Reply:
What I'm leaning towards is that for async2 methods object stack allocation should only stack allocate objects where it can see all field accesses. If all field accesses are visible then stack allocation even across suspension points is ok (they will be captured the same as any other value).

Reply:
Allowing stack allocated objects across suspension points would imply that upon resuming the references to such object would need to be adjusted as the object could be at a different address - we may even be on a different thread/stack. 

I think it is possible to support stackallocated objects with roughly same restrictions as in non-async case, but may not be worth the extra complexity.

Reply:
It would only be allowed for cases where all accesses of the object are visible to the JIT. We would not allow it for cases like the one you describe.